### PR TITLE
Conversion to base map 5/n: Partition

### PR DESCRIPTION
**Pull request by eric-anderson**
* Add new Composite class that models multiple composite operations in a single node
* Use it to rewrite partition which is a sometime composite operation.

### CODE REVIEW

File: lib/sycamore/sycamore/transforms/partition.py [hunk 1]
```diff
@@ -501,23 +502,20 @@ class Partition(Transform):
...
-                **self.resource_args,
-            )
-        else:
-            dataset = input_dataset.map(generate_map_function(self._partitioner.partition))
-        if self._table_extractor:
-            dataset = dataset.map(generate_map_function(self._table_extractor.extract_tables))
-        return dataset
+        ops = []
+
+        if isinstance(partitioner, SycamorePartitioner):
```
Comment (HenryL27) [inline lib/sycamore/sycamore/transforms/partition.py:507]: I hope we can turn this check into a property or method or something but out of scope of this

File: lib/sycamore/sycamore/transforms/base.py [hunk 2]
```diff
@@ -193,3 +193,29 @@ def _update_lineage(cls, from_docs, to_docs):
         to_ids = [d.lineage_id for d in to_docs]
 
         return [MetadataDocument(lineage_links={"from_ids": from_ids, "to_ids": to_ids})]
+
+
+class Composite(UnaryNode):
```
Comment (HenryL27) [inline lib/sycamore/sycamore/transforms/base.py:198]: maybe `CompositeTransform` or `CompositeNode`? idk, just feels a little funny next to `BaseMapTransform` but fine if you wanna keep it

**Review by HenryL27 — APPROVED** [review]
1 minor nit you don't have to pay attention to

File: lib/sycamore/sycamore/transforms/base.py (hunk 2 above)
Comment (eric-anderson) [reply lib/sycamore/sycamore/transforms/base.py:198]: done.

**Pull request merged**